internal/go118_sha1_patch.go

```diff
@@ -0,0 +1,47 @@
+package internal
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	_ "unsafe" // for go:linkname
+
+	goversion "github.com/hashicorp/go-version"
+	"github.com/hashicorp/vault/sdk/version"
+)
+
+const sha1PatchVersionsBefore = "1.12.0"
+
+var patchSha1 sync.Once
+
+//go:linkname debugAllowSHA1 crypto/x509.debugAllowSHA1
+var debugAllowSHA1 bool
+
+// PatchSha1 patches Go 1.18+ to allow certificates with signatures containing SHA-1 hashes to be allowed.
+// It is safe to call this function multiple times.
+// This is necessary to allow Vault 1.10 and 1.11 to work with Go 1.18+ without breaking backwards compatibility
+// with these certificates. See https://go.dev/doc/go1.18#sha1 and
+// https://developer.hashicorp.com/vault/docs/deprecation/faq#q-what-is-the-impact-of-removing-support-for-x-509-certificates-with-signatures-that-use-sha-1
+// for more details.
+// TODO: remove when Vault <=1.11 is no longer supported
+func PatchSha1() {
+	patchSha1.Do(func() {
+		patchBefore, err := goversion.NewSemver(sha1PatchVersionsBefore)
+		if err != nil {
+			panic(err)
+		}
+
+		patch := false
```

Comment:
I would prefer to PR this change directly into the 1.11.x and 1.10.x release branches, and then we don't need any logic about whether to patch.

Reply:
re: putting it directly in 1.11.x vs. 1.10.x, I slightly prefer this way, to keep future backports cleaner. But, changing directly in the 1.11.x and 1.10.x branches is okay, it's just slightly more work, and could create more conflicts.

re: where to put it, the `main.go` makes sense (and I'll move the init there). But, I wanted to keep it available for plugins though, i.e., if someone were to build a built-in plugin to run externally, they should be able to run the patch as well, which would mean it can't *only* be available in the root `main.go`.